client/apis/volumegroupsnapshot/v1alpha1/types.go

```diff
@@ -0,0 +1,311 @@
+	// If the label is added or removed to a volume after a group snapshot
+	// is created, the existing group snapshots won't be modified.
+	// Once a VolumeGroupSnapshotContent is created and the sidecar starts to process
+	// it, the volume list will not change with retries.
+	Selector *metav1.LabelSelector `json:"selector" protobuf:"bytes,1,opt,name=selector"`
+
+	// VolumeGroupSnapshotClassName is the name of the VolumeGroupSnapshotClass
+	// requested by the VolumeGroupSnapshot.
+	// VolumeGroupSnapshotClassName may be left nil to indicate that the default
+	// class will be used.
+	// Empty string is not allowed for this field.
+	// +optional
+	VolumeGroupSnapshotClassName *string `json:"volumeGroupSnapshotClassName,omitempty" protobuf:"bytes,2,opt,name=volumeGroupSnapshotClassName"`
+}
+
+// VolumeGroupSnapshotStatus defines the observed state of VolumeGroupSnapshot
+type VolumeGroupSnapshotStatus struct {
+	// BoundVolumeGroupSnapshotContentName is the name of the VolumeGroupSnapshotContent
+	// object to which this VolumeGroupSnapshot object intends to bind to.
+	// If not specified, it indicates that the VolumeGroupSnapshot object has not
+	// been successfully bound to a VolumeGroupSnapshotContent object yet.
+	// NOTE: To avoid possible security issues, consumers must verify binding between
+	// VolumeGroupSnapshot and VolumeGroupSnapshotContent objects is successful
+	// (by validating that both VolumeGroupSnapshot and VolumeGroupSnapshotContent
+	// point at each other) before using this object.
+	// +optional
+	BoundVolumeGroupSnapshotContentName *string `json:"boundVolumeGroupSnapshotContentName,omitempty" protobuf:"bytes,1,opt,name=boundVolumeGroupSnapshotContentName"`
+
+	// CreationTime is the timestamp when the point-in-time group snapshot is taken
+	// by the underlying storage system.
+	// If not specified, it may indicate that the creation time of the group snapshot
+	// is unknown.
+	// +optional
+	CreationTime *metav1.Time `json:"creationTime,omitempty" protobuf:"bytes,2,opt,name=creationTime"`
+
+	// ReadyToUse indicates if all the individual snapshots in the group are ready
+	// to be used to restore a volume.
+	// If not specified, it means the readiness of a group snapshot is unknown.
+	// +optional
+	ReadyToUse *bool `json:"readyToUse,omitempty" protobuf:"varint,3,opt,name=readyToUse"`
+
+	// Error is the last observed error during group snapshot creation, if any.
+	// This field could be helpful to upper level controllers (i.e., application
+	// controller) to decide whether they should continue on waiting for the group
+	// snapshot to be created based on the type of error reported.
+	// The snapshot controller will keep retrying when an error occurs during the
+	// group snapshot creation. Upon success, this error field will be cleared.
+	// +optional
+	Error *VolumeGroupSnapshotError `json:"error,omitempty" protobuf:"bytes,4,opt,name=error,casttype=VolumeGroupSnapshotError"`
```

Comment:
Can you use the existing VolumeSnapshotError from VolumeSnapshot APIs instead of creating a new VolumeGroupSnapshotError?

Reply:
It can be done, but requires some refactoring and moving `VolumeSnapshotError` to a top level utils.go, which in turn removes from autogenerated functions. You can take a look at the latest commit and let me know if we want to reuse the code.

Reply:
Refactoring to a util didn't work. So I've reused from `VolumeSnapshot` API and imported `client/v6/apis/volumesnapshot/v1` into `VolumeGroupSnapshot`. You can take a look in the latest commit.